src/NHibernate.Test/Async/NHSpecificTest/NH3374/FixtureByCode.cs

```diff
@@ -0,0 +1,119 @@
+// <auto-generated>
+//     This code was generated by AsyncGenerator.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+
+using System.Linq;
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Linq;
+using NHibernate.Mapping.ByCode;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.NH3374
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	[TestFixture]
+	public class ByCodeFixtureAsync : TestCaseMappingByCode
+	{
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<Document>(rc =>
+			{
+				rc.Id(x => x.Id, idMapper => idMapper.Generator(Generators.Identity));
+				rc.ManyToOne(x => x.Blob, m =>
+					{
+						m.Cascade(Mapping.ByCode.Cascade.All);
+					});
+				rc.Property(x => x.Name);
+			});
+			
+			mapper.Class<Blob>(map =>
+				{
+					map.Id(x => x.Id, idMapper => idMapper.Generator(Generators.Identity));
+					map.Property(x => x.Bytes, y =>
+						{
+							y.Length(int.MaxValue);
+							y.Lazy(true);
+						});
+				});
+
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+		protected override void OnSetUp()
+		{
+			using (ISession session = OpenSession())
+			using (ITransaction transaction = session.BeginTransaction())
+			{
+				var e1 = new Document { Name = "Bob" };
+				e1.Blob = new Blob { Bytes = new byte[] { 1, 2, 3 } };
+				session.Save(e1);
+				
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (ISession session = OpenSession())
+			using (ITransaction transaction = session.BeginTransaction())
+			{
+				session.Delete("from System.Object");
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public async Task TestNoTargetExceptionAsync()
+		{
+			Document document = await (LoadDetachedEntityAsync());
+			Blob blob = await (LoadDetachedBlobAsync());
+
+			blob.Bytes = new byte[] { 4, 5, 6 };
+			document.Blob = blob;
+
+			using (ISession session = OpenSession())
+			using (ITransaction transaction = session.BeginTransaction())
+			{
+				session.Merge(document);
+			}
+		}
+
+		private async Task<Blob> LoadDetachedBlobAsync(CancellationToken cancellationToken = default(CancellationToken))
+		{
+			using (ISession session = OpenSession())
+			using (session.BeginTransaction())
+			{
+				var blob = await (session.GetAsync<Blob>(1, cancellationToken));
+				await (NHibernateUtil.InitializeAsync(blob.Bytes, cancellationToken));
+				return blob;
+			}
+		}
+
+		private Task<Document> LoadDetachedEntityAsync(CancellationToken cancellationToken = default(CancellationToken))
+		{
+			try
+			{
+				using (ISession session = OpenSession())
+				using (session.BeginTransaction())
+				{
+					return session.GetAsync<Document>(1, cancellationToken);
```

Comment:
A task should never be directly returned from inside a using statement, otherwise `Dispose` will be called before the returned task is finished. So maybe this is a general bug in the generated code?